### ISSUE TITLE
Example from the front page with constrained expected improvement[Bug]

### ISSUE THREAD

**Issue by aegonwolf** [pytorch/botorch#1497]
# 🐛 Bug
This is probably not a bug but a failure of understanding from my side. I found it hard to find fitting (newbie friendly) examples, so I hope you may excuse me adding this here.
I tried to use the example from the front page but tweaked it for a constrained problem.

## To reproduce

```
import torch
from botorch.models import SingleTaskGP
from botorch.fit import fit_gpytorch_model
from botorch.utils import standardize
from gpytorch.mlls import ExactMarginalLogLikelihood

train_X = torch.rand(10, 2)
Y = 1 - torch.norm(train_X - 0.5, dim=-1, keepdim=True)
Y = Y + 0.1 * torch.randn_like(Y)  # add some noise
train_Y = standardize(Y)

gp = SingleTaskGP(train_X, train_Y)
# mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)

from botorch.acquisition import ConstrainedExpectedImprovement
constraints = {0: (1.2, None)}
cei = ConstrainedExpectedImprovement(mll, 0.2, 1, constraints)

from botorch.optim import optimize_acqf

bounds = torch.stack([torch.zeros(2), torch.ones(2)])
candidate, acq_value = optimize_acqf(
    cei, bounds=bounds, q=1, num_restarts=5, raw_samples=20,
)
candidate  # tensor([0.4887, 0.5063])
```

The error I get is the following:
```
`AttributeError: 'ExactMarginalLogLikelihood' object has no attribute 'posterior'`
```

## Expected Behavior

I get optimized parameters.

**Comment by Balandat** [comment]
The issue is the following line: 

> cei = ConstrainedExpectedImprovement(mll, 0.2, 1, constraints)

Acquisition functions take a `Model` as the first arg, but you are passing in the MLL here. Replacing this with 

> cei = ConstrainedExpectedImprovement(gp, 0.2, 1, constraints)

should work.

**Comment by aegonwolf** [comment]
> The issue is the following line:
> 
> > cei = ConstrainedExpectedImprovement(mll, 0.2, 1, constraints)
> 
> Acquisition functions take a `Model` as the first arg, but you are passing in the MLL here. Replacing this with
> 
> > cei = ConstrainedExpectedImprovement(gp, 0.2, 1, constraints)
> 
> should work.

Thank you! The error has gone, however, it results in a new one from the same function:

`AssertionError: Expected the output shape to match either the t-batch shape of X, or the `model.batch_shape` in the case of acquisition functions using batch models; but got output with shape torch.Size([20, 0]) for X with shape torch.Size([20, 1, 2]).`

**Comment by Balandat** [comment]
The issue is that your model only models one outcome, but `ConstrainedExpectedImprovement` requires the model to have at least two outputs (1 objective and 1+ constraints).

With what you currently have you're solving something like `max f s.t. f >= 1.2` which isn't really something that one would do in practice. 

The code will run (though probably not do anything super useful given the random data) if you add 

```
Y = 1 - torch.norm(train_X - 0.5, dim=-1, keepdim=True)
Y = Y.expand(*Y.shape[:-1], 2)
```
to your data generation to make the model have two outputs (which in this case are identical, except for the noise)

**Comment by aegonwolf** [comment]
> The issue is that your model only models one outcome, but `ConstrainedExpectedImprovement` requires the model to have at least two outputs (1 objective and 1+ constraints).
> 
> With what you currently have you're solving something like `max f s.t. f >= 1.2` which isn't really something that one would do in practice.
> 
> The code will run (though probably not do anything super useful given the random data) if you add
> 
> ```
> Y = 1 - torch.norm(train_X - 0.5, dim=-1, keepdim=True)
> Y = Y.expand(*Y.shape[:-1], 2)
> ```
> 
> to your data generation to make the model have two outputs (which in this case are identical, except for the noise)

I see, thanks a lot for elaborating. 
I didn't plan on it doing anything useful yet, but tried to translate some of the most basic implementations to get a bit more familiar with botorch.

**Comment by aegonwolf** [comment]
As a small side note from a users perspective since I came to botorch after watching an introduction on youtube which sounded great, as the goal seemed to make it simpler.
There is an amazing amount of functionality and you have some great tutorials @Balandat the API is documented reasonably well, however the most basic stuff is missing tutorials (for example stuff you find in bayesopt, skopt, gpyopt etc), so for a non researcher (maybe not the target audience) who is very familiar, these examples seem extremely abstract and far removed from the most basic gp bayesian optimization. I find it hard when I can not relate the tutorials to theory I did understand.

**Comment by Balandat** [comment]
Hi @aegonwolf, thanks for the feedback! While our primary target audience are indeed researchers, we definitely don't want to make things artificially hard for the end user. 

> the most basic stuff is missing tutorials (for example stuff you find in bayesopt, skopt, gpyopt etc)

What specifically are you referring to when you say "the most basic stuff"? Would you mind providing some pointers?

I see most of the packages you mentioned have a simple "fire and forget" loop of the kind `run_the_loop(function, param_defs, num_iter)`. We made a conscious decision not to add a lot of this kind of functionality to BoTorch to make sure that people understand the library is primarily designed for advanced use cases. Exposing all of the functionality we do support in a reasonably clean way in such a one-size-fits-all loop would have been exceedingly hard. For these kinds of situations we point users to Ax instead. 

Ax will auto-select a good strategy and run basic optimization loop cases without much fuss. Now I see you have run into some trouble with hooking a partially custom setup up in Ax - that's unfortunate and we'll help with that. Getting this interface between Ax and BoTorch right to make it easily extensible is quite a challenging problem: On the one side you have a very low-level modular and rather unopinionated research library and on the other side you have a high level interface that should run with minimal code input in a robust fashion while making sure everything is serializable and sendable over the wire. If you happen to have specific feedback how to improve that boundary we'd love to hear that.

**Comment by aegonwolf** [comment]
Hi, thanks a lot for replying.

They do, though I am also meaning basic iterations with custom acquisition functions, which we can do with these packages too.
I.e. 

1. get some initial samples
2. update the gprs
3. get new samples

and in the end find the best samples.

I think I got the gist of it down now, at least for in botorch itself. 
I was not able to do it with Ax, (ps, I think in the tutorial there is an extra cell that specifies ax_model but it is never used, i.e. I was able to run it completely without, which adds some confusion). 
Some things seem super cool, like ModelListGP, i.e. for constraints, etc in the other packages we need to implement separate models manually (they do have that functionality built in to an extent but not very customizable).

So I was excited to find ModelListGP but its mentioned here and there though I was not able to get it to work with ListSurrogate and Ax, and I did not find any examples anywhere. 
Some simple examples on how to use this with Ax would be great.
The Ax Documentation actually offers some more basic tutorials that I was missing from Botorch but also just mentions things here and there without giving examples.

In the end, I was able to use multiple Models but I had to write my own update loop and constrained expected improvement which seems super unnecessary since you have all the parts. I was however not able to put them together.

Should add, writing my own was much easier than I had expected, so the basic botorch API (without ax)  is actually quite intuitive and comparable to the other packages I added (in some ways even easier).

I am sure Ax is probably easy too, I just haven't used either before and from the examples I was not able to figure out how.
